python/ray/train/utils.py

```diff
@@ -31,19 +33,46 @@ def check_for_failure(remote_values: List[ObjectRef]) -> Tuple[bool, List[int]]:
     """
     unfinished = remote_values.copy()
     dead_worker_indexes = []  # Store the indexes of the failed workers.
+
+    at_least_one_failed_worker = False
     while len(unfinished) > 0:
-        finished, unfinished = ray.wait(unfinished)
+        # Once a worker has failed, we add a timeout for getting the
+        # results for the remaining workers.
+        # This is to avoid situations where the remaining workers
+        # are alive, but hanging on collective calls because other
+        # workers have failed.
+        timeout = (
+            REMAINING_WORKERS_GRACE_PERIOD_S if at_least_one_failed_worker else None
```

Comment:
Change `REMAINING_WORKERS_GRACE_PERIOD_S` to support being set as an environment variable? Is 10 seconds too short?